### DISCUSSION TITLE
When to set seeds

### DISCUSSION THREAD

**Discussion by john-harrold** [nlmixr2/rxode2#776] [Q&A]
Hello I was working on some simulations and I was using both `set.seed` and `rxSetSeed` set to the same value. I was calling it right before I called `rxSolve()`, but it didn't seem to take effect (repeated running of the script was giving me different results). However when I placed the seed setting commands before the rxode2 model function definition it seemed to work. Am I imagining this? Or does the seed need to be set before the model function definition to work?

**Comment by mattfidler** [comment]
There is nothing in the model code about random number generation, so I'm unsure what is happening here.

**Comment by mattfidler** [comment]
There are tests for the behavior in `test-seeds.R`  in these cases, the set seed is right before the `rxSolve()` call.

I double checked how `rxSolveSEXP` is called from R/C and it:
Is called from the Rcpp export which means that the internal R seed management is called within the Rcpp macros.  (That is they call `GetRNGstate()` on entering the functon and  `PutRNGstate()` on exiting the macro). (Internal Rcpp code starts here: https://github.com/RcppCore/Rcpp/blob/1e82f1872eebe4fc813161ec0bac9bdfc53e1619/src/api.cpp#L56)
All R seed management works OK.

When `rxSetSeed()` isn't called, then it uses a random number from R to seed the simulation (which is not ideal). 

Each solve uses independent seed based on the initial seed selected + the simulation id number.  Afterwards it increments the initial seed by the number of simulation subjects.

The only issue I noticed reviewing the code is this is not used in dop853, it calls the rxode2 seed in whatever state it is in.  

Were you using `lsoda` followed by `dop853`?

If so perhaps that is the issue. 

A reproducible example would really help here.

**Comment by john-harrold** [comment]
I apologize. I'm not sure what I was doing wrong before but I moved it right after the model definition and it's working now. Sorry.

**Reply by mattfidler** [reply]
Don't worry about  it :smile:
